### PR TITLE
fix: fix alter table verification

### LINKED ISSUE #2425: Table broken after running incorrect alter statement (standalone mode) ?
### What type of bug is this?

Unexpected error

### What subsystems are affected?

Standalone mode (latest commit: e7e254cd1164df8fe1052add8ce)

### What happened?

Table is broken after running incorrect alter statement. (standalone mode)

You can reproduce this issue by following steps:

1. Create a table like this:
```sql
create table table_name(value_double double not null, ts TIMESTAMP DEFAULT CURRENT_TIMESTAMP TIME INDEX);
```

2. Run incorrect alter table:
```sql
alter table table_name add column host string not null;
```

This should fail since we can't add a new_column which doesn't allow null and have default value. So, this is no problem.

Once we run the above steps, we can't execute select statements like `select * from table_name`. (maybe any statement for this table)

You can any statements which aren't related to this table, like so:
```sql
show databases;
```

### What operating system did you use?

Mac OS (Ventura 13.3) (Maybe any OS)

### Relevant log output and stack trace

```bash
Here's a history of mysql client.


mysql> create table table_name(value_double double not null, ts TIMESTAMP DEFAULT CURRENT_TIMESTAMP TIME INDEX);
Query OK, 0 rows affected (0.01 sec)

mysql> select * from table_name;
Empty set (0.00 sec)

mysql> alter table table_name add column host string not null;
ERROR 1815 (HY000): Failed to execute query, source: Failed to execute ddl, source: Failed to wait procedure done, source: Failed to execute procedure, source: Failed to execute procedure due to external error, source: Failed to operate on datanode: peer-0(), source: Failed to handle request for region 4402341478400(1025, 0), source: Invalid region request, region_id: 4402341478400(1025, 0), err: no default value for column host, location: src/store-api/src/region_request.rs:323:9, location: src/mito2/src/worker/handle_alter.
mysql> select * from table_name;
ERROR 2013 (HY000): Lost connection to MySQL server during query
No connection. Trying to reconnect...
Connection id:    8
Current database: public

ERROR 2013 (HY000): Lost connection to MySQL server during query
No connection. Trying to reconnect...
Connection id:    8
Current database: public

ERROR 2013 (HY000): Lost connection to MySQL server during query
```

By the way, after closing mysql client, it seems like there's the same problem.

### How can we reproduce the bug?

1. Create a table like this:
```sql
create table table_name(value_double double not null, ts TIMESTAMP DEFAULT CURRENT_TIMESTAMP TIME INDEX);
```

2. Run incorrect alter table:
```sql
alter table table_name add column host string not null;
```

This should fail since we can't add a new_column which doesn't allow null and have default value.

Once we run the above steps, we can't execute select statements like `select * from table_name`. (maybe any statement for this table)

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

1. fix alter table verification
2.  move AlterTable UpdateMetadata to last step

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
fix #2425

### CODE REVIEW

File: src/table/src/metadata.rs [hunk 1]
```diff
@@ -248,6 +248,41 @@ impl TableMeta {
...
+                !names.contains(&col_to_add.column_schema.name),
+                error::InvalidAlterRequestSnafu {
+                    table: table_name,
+                    err: format!(
+                        "add column {} more than once",
+                        col_to_add.column_schema.name
+                    ),
+                }
+            );
+            names.insert(&col_to_add.column_schema.name);
```
Comment (MichaelScofield) [inline src/table/src/metadata.rs:264]:
```suggestion
            ensure!(
                names.insert(&col_to_add.column_schema.name),
                error::InvalidAlterRequestSnafu {
                    table: table_name,
                    err: format!(
                        "add column {} more than once",
                        col_to_add.column_schema.name
                    ),
                }
            );
```

**Review by MichaelScofield — APPROVED** [review]

File: src/meta-srv/src/procedure/tests.rs [hunk 2]
```diff
@@ -393,7 +393,7 @@ async fn test_submit_alter_region_requests() {
     });
 
     let status = procedure.submit_alter_region_requests().await.unwrap();
-    assert!(matches!(status, Status::Done));
+    assert!(matches!(status, Status::Executing { persist: true }));
 
     handle.await.unwrap();
```
Comment (NiwakaDev) [inline src/meta-srv/src/procedure/tests.rs:399]: I guess that we need to add not only this test but also a integration test.
```sql
create table ~ 
alter table ~ add column foo string not null 
select * from ~
```

Comment (WenyXu) [reply src/meta-srv/src/procedure/tests.rs:399]: Let’s add these tests once the `sqlness` test is available.

File: src/table/src/metadata.rs [hunk 3]
```diff
@@ -248,6 +248,40 @@ impl TableMeta {
...
+                }
+            );
+
+            ensure!(
+                !table_schema.contains_column(&col_to_add.column_schema.name),
+                error::ColumnExistsSnafu {
+                    table_name,
+                    column_name: col_to_add.column_schema.name.to_string()
+                },
+            );
```
Comment (fengjiachun) [inline src/table/src/metadata.rs:271]:
```suggestion
            );

```

**Review by fengjiachun — APPROVED** [review]
LGTM

File: src/meta-srv/src/procedure/tests.rs (hunk 2 above)
Comment (evenyag) [reply src/meta-srv/src/procedure/tests.rs:399]: @WenyXu `sqlness` test is available now, do you have time to add this test?

Comment (NiwakaDev) [reply src/meta-srv/src/procedure/tests.rs:399]: @evenyag  @WenyXu 
I'd like to work on this if you don't mind.

Comment (WenyXu) [reply src/meta-srv/src/procedure/tests.rs:399]: @NiwakaDev Thanks 🙏

**Pull request merged**